doc/developer/design/20230531_compute_metrics.md

```diff
@@ -0,0 +1,414 @@
+Database components can register and update metrics to be exposed through these endpoints using the [`MetricsRegistry`] type.
+
+[`MetricsRegistry`]: https://github.com/MaterializeInc/materialize/blob/244bdc6cc6c6180cc1b92e1b84915dcaa52685e6/src/ore/src/metrics.rs#L105
+
+### prometheus-exporter
+
+The prometheus-exporter is a service running in Materialize cloud environments that generates Prometheus metrics from SQL query results.
+It is configured with a set of queries to run at regular intervals.
+The results of these queries are converted into metric labels and values and then exposed through a metric endpoint.
+
+The prometheus-exporter is capable of performing per-replica queries, making it suitable for collecting metrics derived from introspection sources.
+
+Note that for prometheus-exporter metrics we normally use the gauge type even for metrics with strictly increasing values.
+That is because we replace the metric value every time with the query result, which the counter type doesn't allow (it can only be increased by a relative amount).
+It would be possible to adapt the prometheus-exporter to also support the counter type, by remembering previous values, but so far that has not been necessary.
+
+## Compute Metrics
+
+In this document, we discuss two groups of metrics:
+
+* Controller metrics, i.e. metrics exported by the compute controller
+* Replica metrics, i.e. metrics exported by compute replicas
+
+Controller metrics are always exposed through direct export.
+
+Replicas are able to write to introspection sources, so replica metrics can be exposed either through direct export or through the prometheus-exporter querying introspection sources.
+In general, we prefer exporting replica information through introspection sources, as this way the information also becomes useful to database users.
+Replica metrics are exported directly when they are unlikely to be useful to database users, for example when they track implementation details of Compute.
+
+### Controller Metrics
+
+The following list describes the metrics we want to collect in the controller.
+All metrics in this list are exposed through direct export using the controller's existing `MetricsRegistry`.
+
+All metrics in this list have an `instance_id` label identifying the compute instance and, if applicable, a `replica_id` label.
+
+* Compute protocol
+  * `mz_compute_commands_total`
+    * **Type**: counter
+    * **Labels**: `instance_id`, `replica_id`, `type`
+    * **Description**: The total number of compute commands sent, by replica and command type.
+  * `mz_compute_responses_total`
+    * **Type**: counter
+    * **Labels**: `instance_id`, `replica_id`, `type`
+    * **Description**: The total number of compute responses sent, by replica and response type.
+  * `mz_compute_command_message_size_bytes`
+    * **Type**: histogram
```

Comment:
The main issue with histogram is that they add a bunch of per-bucket time series behind the scenes, right? My stance in this design was that it's probably fine to add these as long as the metric is otherwise low-cardinality. We definitely want to avoid histograms with a `collection_id` label.

But I agree we shouldn't add unnecessary histograms. Having a counter for the message sizes sounds reasonable to me, I'll change that.

What do you mean by consolidating the result label? Omitting it? My thinking was that the result would allow us to see how long it takes for us to respond to a peek, but also how long people are willing to wait before they cancel a peek. So what I really care about is whether or not the peek was cancelled, not so much whether it returned successfully or with an error.

Reply:
> The main issue with histogram is that they add a bunch of per-bucket time series behind the scenes, right? My stance in this design was that it's probably fine to add these as long as the metric is otherwise low-cardinality.

yup. and that's reasonable. was mostly an fyi that persist started from an even more extreme place of not having any until they proved necessary

> What do you mean by consolidating the result label? Omitting it? My thinking was that the result would allow us to see how long it takes for us to respond to a peek, but also how long people are willing to wait before they cancel a peek. So what I really care about is whether or not the peek was cancelled, not so much whether it returned successfully or with an error.

yeah, sorry, I didn't have enough domain knowledge to have a specific counter-proposal. given this context, I'd say the strawman would be to either discard "error" timings or fold "error" into "success". if you don't care about per-instance_id breakdowns of how long people are willing to wait before cancelling, then you could split cancels into a second histogram which doesn't even have the instance-id label

(to be clear, all the comments I've made on this doc are just suggestions, feel free to ignore any/all of them)